### DISCUSSION TITLE
How to code a simple modbus server programm ?

### DISCUSSION THREAD

**Discussion by varionix** [eModbus/eModbus#369] [Q&A] [answered]
how do i program a modbus server with the esp32-s3-devkitm-1 with vscode and platformio. The library at "https://github.com/eModbus/eModbus.git" is to be used for this.
A half-duplex RS-485 transceiver is used without DT/RE switching.
The server should have the following registers:
5 holding registers of data type signed long integer which are read and written
2 input registers of data type unsigned long integer which are read only
The Modbus client should also be able to access selected holding or input registers of the server (with start register, number of registers to be read).

**Comment by Miq1** [comment]
You may start with the [RTUServer example](https://github.com/eModbus/eModbus/tree/master/examples/RTUserver), try to understand it - may be even consider the extensive [documentation](http://emodbus.github.io/).
You are welcome to post your code here to have it examined and to get help of course.

**Reply by varionix** [reply]
I used this example and connected QModMaster to receive the server answer from this example. 
In QModMaster set baudrate to 19200 (like the example) checked the RX, TX pins, Set start adress to 2 ...get timeout error

**Reply by Miq1** [reply]
There are a couple of common reasons why the RTU communication may fail:
- wiring. UART pins switched, cable broken, RE/DE connection to wrong GPIO. RS485's A+ crossed to B- instead of A->A and B->B
- baud rates, data and stop bit settings
- missing termination resistors (120 Ohms between A+ and B-)
- missing RS485 biasing resistors (4K7 at *one* end between A+ and Vcc, another between B- and GND)
- wrong server ID in request

Plus a number of more esoteric causes.

**Reply by Miq1** [reply]
Did you see progress?

**Reply by varionix** [reply]
No i do not have any progress . i have doublechecked all wirings, put a 120 ohm terminator.
i get on the terminal the error : ModbusServerRTU.cpp  [ 259] serve: RTU receive: E5 - Packet length error

**Reply by Miq1** [reply]
That at least proves something is going back and forth. Please add a ``build_flags = -DLOG_LEVEL=6`` to your ``platformio.ini``, compile again and post the log output here to have it analyzed.

The E5 error is pointing to a problem with the RE/DE line handling. We had a couple of cases where the switch from listening to writing would generate stray zero bytes on the UART, most probably caused by a floating pin during the switch.

**Reply by varionix** [reply]
It is working now. I had an bad connected GDN Pin.
Now i will develop my Modbus Application. Questions left.
Can i use a modbus TPU and modbus TCP together within the same server program. 
Can i use the library to make Client with c++ and not using an ESP32. Will use QT

**Reply by Miq1** [reply, marked as answer]
> It is working now. I had an bad connected GDN Pin. Now i will develop my Modbus Application. 

Excellent!

> Questions left. Can i use a modbus TPU and modbus TCP together within the same server program. 

Yes.

> Can i use the library to make Client with c++ and not using an ESP32. Will use QT

Yes, provided you are using Linux. There is a Linux client adaption to be found at ``examples/Linux`` that will allow that. Please be sure to read the Readme docs there, as you will need to copy some library files first, build the static ``libeModbus`` and install it. There are ``Makefile``s to support that.

**Reply by varionix** [reply]
MODBUS function code 08 (0x08) Diagnostics.
can i implement this function in my server programm and how to programm.

**Reply by Miq1** [reply]
You will want to write a worker function that will provide the requested data as a response and register that function for function code 0x08 - just like it is done in the server examples, there mostly for function code 0x03 for demonstration purposes.

**Comment by varionix** [comment]
i would like to extend the innofficial Error Codes like below:

0xC0 - BUFFER OVERFLOW:   The internal buffer has overflowed, which means that more data has been received than can be processed.
0xC1 - INVALID RESPONSE LENGTH: The length of the received response is invalid or does not match expectations.
0xC2 - INVALID UNIT ID: The unit ID received is invalid or is not supported.
0xC3 - INVALID COMMAND FORMAT: The format of the command received is invalid or incorrect.
0xC4 - MEMORY ACCESS ERROR: An error has occurred while accessing the device's memory.
0xC5 - PROTOCOL ERROR: A general protocol error that indicates a communication error that is not covered by more specific error codes.
0xC6 - BAD GATEWAY: Error routing the request through a gateway, possibly due to a faulty configuration or a communication problem.
0xC7 - CONNECTION LOST: The connection to the server (slave) was unexpectedly interrupted.
0xC8 - TRANSACTION ID MISMATCH: The transaction ID of the response does not match that of the request, which indicates a problem with the synchronisation.
0xC9 - INVALID STARTING ADDRESS: The specified starting address is invalid or not in the address range of the device.
0xCA - FRAME SEQUENCE ERROR: An error in the frame sequence was detected, indicating a corrupted transmission.
0xCB - DUPLICATE PACKET: A duplicate packet was detected, possibly due to a fault in the communication channel.
0xCC - MALFORMED PACKET: A faulty or incomplete packet has been received that cannot be processed.

i know this have to be added in ModbusTypeDefs.h but where to implement else in the library . I assume in the ModbusMessage.h

**Reply by Miq1** [reply]
I do not like that to be added to the "official" code, as the Modbus standard does not specify too much on function code 0x08. Even our internal error codes are already irritating som devices out in the wild obviously, so I have an option already to suppress those on my to-do list.
But you are of course free to add that to your local fork!
